### PR TITLE
18 - Add books by entering title / isbn

### LINKED ISSUE #18: Add books by entering title/ISBN
Allow users to enter the ISBN or the title of the book in a dialog.

### PR DESCRIPTION
**Pull request by shockbytes**
Implement bottom sheet for adding new books from a query string. Furthermore, this PR applies a couple of small bug fixes in the repository layer.

NOTE: The UI is not final in the bottom sheet but it allows us to add books. This is a prerequisite for virtually all other tickets.

Sorry for the german screenshots

| Enter query | Bottom sheet | Book added |
|-|-|-|
|Simulator Screen Shot - iPhone 14 Pro Max - 2022-11-19 at 22 15 14|Simulator Screen Shot - iPhone 14 Pro Max - 2022-11-19 at 22 15 44|Simulator Screen Shot - iPhone 14 Pro Max - 2022-11-19 at 22 15 48|

### CODE REVIEW

File: lib/l10n/app_de.arb [hunk 1]
```diff
@@ -3,8 +3,13 @@
...
   "add_scan": "Buchcode scannen",
   "add_query": "Titelsuche",
-  "add_manual": "Manuell eingeben"
-}
\ No newline at end of file
+  "add_manual": "Manuell eingeben",
+  "cancel": "Zurück",
+  "search": "Suchen",
+  "query_search_hint": "Titel oder ISBN",
+  "query_search_title": "Titelsuche"}
```
Comment (lockieRichter) [inline lib/l10n/app_de.arb]: Nit: JSON formatting here, might be useful to use a plugin for formatting JSON standard and also keeping in alphabetical order.

File: lib/src/bloc/add/add_book_bloc.dart [hunk 2]
```diff
@@ -1,12 +1,43 @@
+import 'package:dantex/src/data/book/book_repository.dart';
+import 'package:dantex/src/data/book/entity/book.dart';
+import 'package:dantex/src/data/book/entity/book_state.dart';
 import 'package:dantex/src/data/bookdownload/book_downloader.dart';
 import 'package:dantex/src/data/bookdownload/entity/book_suggestion.dart';
+import 'package:rxdart/rxdart.dart';
 
 class AddBookBloc {
```
Comment (lockieRichter) [inline lib/src/bloc/add/add_book_bloc.dart:8]: Can you write some tests for this class please? I will try to setup a GitHub actions in my next change that runs the tests so we can prevent regressions!

File: lib/src/core/injection/dependency_injector.dart [hunk 3]
```diff
@@ -6,13 +6,22 @@ import 'package:get/get.dart';
 
 import 'blocking_components_injector.dart';
 
+/// Nice little wrapper for Dependency Injection code.
+/// Allows to switch the underlying dependency injection library, if needed.
 class DependencyInjector {
```
Comment (lockieRichter) [inline lib/src/core/injection/dependency_injector.dart:11]: Nice one, love this.

File: lib/src/ui/add/add_book_sheet.dart [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+class AddBookSheet extends StatefulWidget {
+  final String _query;
+
+  const AddBookSheet(
+    this._query, {
+    Key? key,
+  }) : super(key: key);
+
+  @override
+  State<StatefulWidget> createState() => _AddBookSheetState();
```
Comment (lockieRichter) [inline lib/src/ui/add/add_book_sheet.dart]: I typically don't include this in my state declarations as it's not needed. I think we either should **always** include it or **never** include it for consistency. 

Happy to go with whatever you want.

**Review by lockieRichter — COMMENTED** [review]
Looks really good, couple of minor comments!

File: lib/l10n/app_de.arb (hunk 1 above)
Comment (shockbytes) [reply lib/l10n/app_de.arb]: Thanks, I overlooked this one.

File: lib/src/ui/add/add_book_sheet.dart (hunk 4 above)
Comment (shockbytes) [reply lib/src/ui/add/add_book_sheet.dart]: I am usually a proponent of complete type declarations. However, `createState` is a well-known method, let's go with the short version.

**Review by lockieRichter — COMMENTED** [review]
Everything looking good, just one unresolved thread!

**Comment by shockbytes** [comment]
Pushed the test cases for now. They already caught 2 bugs ;)

**Review by lockieRichter — APPROVED** [review]
Great work

**Pull request merged**